src/gs-update-monitor.c

```diff
@@ -150,6 +190,269 @@ no_updates_for_a_week (GsUpdateMonitor *monitor)
+		}
+		return;
+	}
+
+	if (state == AS_APP_STATE_INSTALLING) {
+		/* if we cannot update at the moment, cancel any automatically
+		 * started update */
+		if (!app_is_auto_updating (helper->app))
+			return;
+
+		g_debug ("Cancelling automatic update of app %s, as "
+			 "download-now is FALSE", app_id);
+		g_cancellable_cancel (gs_app_get_cancellable (helper->app));
+	}
+}
+
+static void
+monitor_clear_scheduler (GsUpdateMonitor *monitor)
+{
+	g_clear_object (&monitor->scheduler);
+}
+
+static void
+async_result_cb (GObject *source_object,
+                 GAsyncResult *result,
+                 GAsyncResult **out_result)
+{
+	*out_result = g_object_ref (result);
+}
+
+static gboolean
+setup_scheduler (GsUpdateMonitor *monitor,
+		 GCancellable *cancellable,
+		 GError **error)
+{
+	g_autoptr(GAsyncResult) new_result = NULL;
+	GMainContext *context = NULL;
+	MwscScheduler *scheduler = NULL;
+
+	if (monitor->scheduler != NULL)
+		return TRUE;
+
+	context = g_main_context_default ();
+
+	mwsc_scheduler_new_async (cancellable,
+				  (GAsyncReadyCallback) async_result_cb,
+				  &new_result);
+	while (new_result == NULL)
+		g_main_context_iteration (context, TRUE);
+
+	scheduler = mwsc_scheduler_new_finish (new_result, error);
+
+	if (scheduler == NULL)
+		return FALSE;
+
+	g_signal_connect_swapped (scheduler, "invalidated",
+				  (GCallback) monitor_clear_scheduler,
+				  monitor);
+	monitor->scheduler = scheduler;
+
+	return TRUE;
+}
+
+static void
+scheduled_entry_invalidated_cb (MwscScheduleEntry *entry,
+				const GError *error,
+				UpdateScheduleHelper *helper)
+
+{
+	const gchar *app_id = gs_app_get_unique_id (helper->app);
+
+	g_debug ("Removing scheduled update of app %s", app_id);
+	g_hash_table_remove (helper->monitor->scheduled_updates, app_id);
+}
+
+static void
+schedule_update (GsUpdateMonitor *monitor,
+		 GsApp *app,
+		 GCancellable *cancellable)
+{
+	g_auto(GVariantDict) parameters_dict = G_VARIANT_DICT_INIT (NULL);
+	g_autoptr(GVariant) parameters = NULL;
+	g_autoptr(GError) local_error = NULL;
+	g_autoptr(GAsyncResult) new_result = NULL;
+	UpdateScheduleHelper *helper = NULL;
+	GMainContext *context = NULL;
+	const gchar *app_id = gs_app_get_unique_id (app);
+
+	helper = g_hash_table_lookup (monitor->scheduled_updates, app_id);
+	if (helper != NULL) {
+		/* replace the app that's scheduled, in case the object is
+		 * different */
+		if (helper->app != app) {
+			GCancellable *app_cancellable = gs_app_get_cancellable (helper->app);
+			g_cancellable_cancel (app_cancellable);
+			g_set_object (&helper->app, app);
+		}
+
+		return;
+	}
+
+	g_variant_dict_insert (&parameters_dict, "resumable", "b", FALSE);
+	parameters = g_variant_ref_sink (g_variant_dict_end (&parameters_dict));
+
+	if (!setup_scheduler (monitor, cancellable, &local_error)) {
+		g_warning ("Could not schedule update for app %s: %s",
+			   app_id, local_error->message);
+		return;
+	}
+
+	if (helper == NULL) {
+		helper = g_new0 (UpdateScheduleHelper, 1);
```

Comment:
I'd use `g_slice_new0` here, but maybe my knowledge is outdated and it's not recommended anymore?

Reply:
There’s not much of a performance difference any more. I use `g_new0()` because the API is a little nicer.

Reply:
@mariospr I think it's no longer an advantage to use the slice API, that's why I've used the normal `g_new0`.